glean-core/src/metrics/timing_distribution.rs

```diff
@@ -171,41 +171,90 @@ impl TimingDistributionMetric {
+    /// implementing class is a [TimingDistributionMetricType] and the instance this
+    /// method was called on is using [TimeUnit.Second], then `samples` are assumed
+    /// to be in that unit).
+    ///
+    /// ## Arguments
+    ///
+    /// - `samples` - The vector holding the samples to be recorded by the metric.
+    pub fn accumulate_samples(&mut self, glean: &Glean, samples: Vec<u32>) {
         if !self.should_record(glean) {
             return;
         }
 
-        glean
-            .storage()
-            .record_with(&self.meta, |old_value| match old_value {
-                Some(Metric::TimingDistribution(mut hist, time_unit)) => {
-                    hist.accumulate(duration as u32);
-                    Metric::TimingDistribution(hist, time_unit)
-                }
-                _ => {
-                    let mut hist = Histogram::default();
-                    hist.accumulate(duration as u32);
-                    Metric::TimingDistribution(hist, self.time_unit)
-                }
-            });
+        for sample in samples {
+            glean
+                .storage()
+                .record_with(&self.meta, |old_value| match old_value {
+                    Some(Metric::TimingDistribution(mut hist, time_unit)) => {
+                        hist.accumulate(sample);
+                        Metric::TimingDistribution(hist, time_unit)
+                    }
+                    _ => {
+                        let mut hist = Histogram::default();
+                        hist.accumulate(sample);
+                        Metric::TimingDistribution(hist, self.time_unit)
+                    }
+                });
+        }
     }
 
-    /// Abort a previous [start] call. No error is recorded if no [start] was called.
+    /// Accumulates the provided signed samples in the metric.
     ///
-    /// @param timerId The [GleanTimerId] to associate with this timing. This allows
-    /// for concurrent timing of events associated with different ids to the
-    /// same timing distribution metric.
+    /// This is required so that the platform-specific code can provide us with
+    /// 64 bit signed integers if no `u32` comparable type is available. This
+    /// will take care of filtering and reporting errors for any provided negative
+    /// sample.
     ///
-    pub fn cancel(&mut self, id: TimerId) {
-        self.timings.cancel(id);
+    /// ## Arguments
+    ///
+    /// - `samples` - The vector holding the samples to be recorded by the metric.
+    ///
+    /// ## Notes
+    ///
+    /// Discards any negative value in `samples` and report an `ErrorType::InvalidValue`
+    /// for each of them.
+    pub fn accumulate_samples_signed(&mut self, glean: &Glean, samples: Vec<i64>) {
+        let sample_count = samples.len();
+        let positive_samples: Vec<u32> = samples
+            .into_iter()
+            .filter(|s| *s >= 0)
+            .map(|s| s as u32)
+            .collect();
```

Comment:
How big do we expect these vectors to be?  If large, maybe it's an issue that it does a copy...?  (Alternatively, `accumulate_samples` could take an iterator and avoid the copy, but then the negative detection would have to happen in a slightly different way...)

Reply:
The size of the vectors is to be defined: this is a potential hot path though, so anything we could do here to optimize might be a good win.

Reply:
IMO we should simplify this and even avoid rapidly calling record_with all the time.
Just use a for loop, do an if check in there and only accumulate if it is positive.

Reply:
```rust
let mut num_errors = 0;
glean
.storage()
    .record_with(&self.meta, |old_value| {
        let (mut hist, time_unit) = match old_value {
            Some(Metric::TimingDistribution(mut hist, time_unit)) => {
                (hist, time_unit)
            }
            _ => {
                (Histogram::default(), self.time_unit);
            }
        };

        for sample in samples { 
            if sample >= 0 {
                hist.accumulate(sample);
            } else {
                num_errors += 1;
            }
        }


        Metric::TimingDistribution(hist, time_unit)
    });    

let msg = format!("Accumulated {} negative samples", num_errors);
record_error(glean, &self.meta, ErrorType::InvalidValue, msg, num_errors);
```

Reply:
Thanks. I had to slightly tweak this because `samples` was not getting captured. Using `move` created another set of problems, which I was not able to deal with :)